### PR TITLE
HIVE-25086: Create Ranger Deny Policy for replication db in all cases.

### PR DESCRIPTION
**Pull request by hmangla98**
Create Ranger Deny Policy for replication db in all cases if hive.repl.ranger.target.deny.policy is set to true.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 1]
```diff
@@ -127,8 +127,9 @@ public int execute() {
       if (shouldLoadAtlasMetadata()) {
         addAtlasLoadTask();
       }
-      if (shouldLoadAuthorizationMetadata()) {
-        initiateAuthorizationLoadTask();
+      if (conf.getBoolVar(HiveConf.ConfVars.REPL_INCLUDE_AUTHORIZATION_METADATA)
+              || conf.getBoolVar(HiveConf.ConfVars.REPL_RANGER_ADD_DENY_POLICY_TARGET)) {
+        initiateRangerLoadTask();
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:132]: better to create a separate task for adding deny policy

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 2]
```diff
@@ -160,14 +161,13 @@ public int execute() {
     }
   }
 
-  private boolean shouldLoadAuthorizationMetadata() {
-    return conf.getBoolVar(HiveConf.ConfVars.REPL_INCLUDE_AUTHORIZATION_METADATA);
-  }
-
-  private void initiateAuthorizationLoadTask() throws SemanticException {
+  private void initiateRangerLoadTask() throws SemanticException {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:164]: this will also try and replicate ranger policies unnecessarily. Just adding a deny policy should be sufficient.

Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:164]: if REPL_INCLUDE_AUTHORISATION_METADATA is not enabled, rangerLoadRoot would be set to null and RangerLoadTask would not replicate ranger policies if rangerLoadRoot is null. It would only create a deny policy for target db in this case.

**Review by aasha — COMMENTED** [review]
Did you generate the thrift files using the maven command?

**Comment by aasha** [comment]
What happens if the config is first enabled and then disable? Do we support it?

**Comment by hmangla98** [comment]
> Did you generate the thrift files using the maven command?

Yes

**Comment by hmangla98** [comment]
> What happens if the config is first enabled and then disable? Do we support it?

Currently, user would have to manually disable the ranger deny policy for target db if config is disabled afterwards.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/RangerDenyTask.java [hunk 3]
```diff
@@ -131,6 +141,23 @@ public int execute() {
         }
     }
 
+    private boolean isSetRangerDenyPolicyForReplicatedDb(String rangerEndpoint, String rangerHiveServiceName,
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/RangerDenyTask.java:144]: this can be expensive. Is there any other alternative to exporting all the ranger policies

Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/RangerDenyTask.java:144]: One way can be save this information in dump dir so that next iteration can know ranger deny policy was already created in previous iteration.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 4]
```diff
@@ -127,6 +127,7 @@ public int execute() {
       if (shouldLoadAtlasMetadata()) {
         addAtlasLoadTask();
       }
+      initiateRangerDenytask();
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:130]: you can do this only if deny config is enabled

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 5]
```diff
@@ -79,6 +74,7 @@
   private static final String RANGER_REST_URL_EXPORTJSONFILE = "service/plugins/policies/exportJson";
   private static final String RANGER_REST_URL_IMPORTJSONFILE =
       "service/plugins/policies/importPoliciesFromFile";
+  private static final String RANGER_REST_URL_DELETEPOLICY = "service/public/v2/api/policy";
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:77]: is this not available as part of service/plugins/policies?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 6]
```diff
@@ -444,7 +482,7 @@ boolean checkConnectionPlain(String url, HiveConf hiveConf) {
     List<RangerPolicy.RangerPolicyItemAccess> denyExceptionsPolicyItemAccesses = new ArrayList<RangerPolicy.
         RangerPolicyItemAccess>();
 
-    resourceNameList.add(sourceDb);
+    resourceNameList.add(targetDb);
     resourceNameList.add("dummy");
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:486]: is this dummy needed?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 7]
```diff
@@ -420,8 +458,8 @@ boolean checkConnectionPlain(String url, HiveConf hiveConf) {
   }
 
   @Override
-  public List<RangerPolicy> addDenyPolicies(List<RangerPolicy> rangerPolicies, String rangerServiceName,
-                                            String sourceDb, String targetDb) throws SemanticException {
+  public RangerPolicy getDenyPolicyForReplicatedDb(String rangerServiceName,
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:461]: are u using the source db param

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java (hunk 6 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:486]: It is required to avoid new deny policy from overriding the existing policy. Ref. HIVE-24371.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java (hunk 7 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:461]: yes, while defining ranger policy name.

Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:461]: Yes, it is being used while specifying the policy  name.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java (hunk 4 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:130]: If config is disabled, it would make delete request if deny policy was created in previous iteration.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java (hunk 5 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:77]: No, This method is not available.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java (hunk 4 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:130]: If config is enabled, this task would add the ranger deny policy.
If config is disabled, this task would delete ranger deny policy if created previously.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 8]
```diff
@@ -127,6 +127,9 @@ public int execute() {
       if (shouldLoadAtlasMetadata()) {
         addAtlasLoadTask();
       }
+      if (conf.getBoolVar(HiveConf.ConfVars.REPL_RANGER_HANDLE_DENY_POLICY_TARGET)) {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:130]: if this conf is set to false but ranger replication is enabled what do you do

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java (hunk 5 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:77]: Please cross check with Ranger team/public docs

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 9]
```diff
@@ -172,6 +168,40 @@ public String getRangerExportUrl(String sourceRangerEndpoint, String rangerHiveS
...
+            case HttpServletResponse.SC_UNAUTHORIZED:
+              throw new Exception("Authentication Failure while communicating to Ranger admin");
+            default:
+              throw new Exception("Ranger policy deletion failed, Please refer target Ranger admin logs.");
+          }
+        }
+        return null;
+      });
+    } catch (Exception e) {
+      throw new SemanticException(ErrorMsg.REPL_RETRY_EXHAUSTED.format(e.getMessage()), e);
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:201]: if its unauthorised even retry will not help
what about any other error

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 10]
```diff
@@ -444,7 +482,7 @@ boolean checkConnectionPlain(String url, HiveConf hiveConf) {
     List<RangerPolicy.RangerPolicyItemAccess> denyExceptionsPolicyItemAccesses = new ArrayList<RangerPolicy.
         RangerPolicyItemAccess>();
 
-    resourceNameList.add(sourceDb);
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:447]: why is the source db needed?

Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:447]: It is included in policy name.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java (hunk 8 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:130]: ranger deny task will not be initiated.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java (hunk 5 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:77]: ok

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/RangerDenyTask.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            if (rangerRestClient == null) {
+                rangerRestClient = getRangerRestClient();
+            }
+            URL url = work.getRangerConfigResource();
+            if (url == null) {
+                throw new SemanticException(ErrorMsg.REPL_INVALID_CONFIG_FOR_SERVICE
+                        .format("Ranger configuration is not valid "
+                                + ReplUtils.RANGER_CONFIGURATION_RESOURCE_NAME, ReplUtils.REPL_RANGER_SERVICE));
+            }
+            conf.addResource(url);
```
[the remark this answers is not in the document]
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/RangerDenyTask.java:88]: This will add existing ranger resource config. Properties of this resource will be override existing properties in conf.

File: common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java [hunk 12]
```diff
@@ -627,7 +627,9 @@
   REPL_INVALID_ARGUMENTS(40012, "Invalid arguments error : {0}.", true),
   REPL_INVALID_ALTER_TABLE(40013, "{0}Unable to alter table{1}", true),
   REPL_PERMISSION_DENIED(40014, "{0}org.apache.hadoop.security.AccessControlException{1}", true),
-  REPL_DISTCP_SNAPSHOT_EXCEPTION(40015, "SNAPSHOT_ERROR", true)
+  REPL_DISTCP_SNAPSHOT_EXCEPTION(40015, "SNAPSHOT_ERROR", true),
+  RANGER_AUTHORISATION_FAILED(40016, "Authorisation Failure while communicating to Ranger admin", true),
```
Comment (aasha) [inline common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:631]: type : authorization

Comment (aasha) [inline common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:631]: error code is same. it should be different. Check if any of the existing errror codes that can be reused.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 13]
```diff
@@ -90,11 +86,13 @@ public RangerExportPolicyList exportRangerPolicies(String sourceRangerEndpoint,
         "cannot be empty", ReplUtils.REPL_RANGER_SERVICE));
     }
     Retryable retryable = Retryable.builder()
-      .withHiveConf(hiveConf)
-      .withRetryOnException(Exception.class).build();
+      .withHiveConf(hiveConf).withFailOnException(RuntimeException.class)
+      .withRetryOnException(URISyntaxException.class).build();
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:90]: what is the use of retry in case of URISyntaxException?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 14]
```diff
@@ -202,12 +238,14 @@ public RangerExportPolicyList importRangerPolicies(RangerExportPolicyList ranger
...
     Retryable retryable = Retryable.builder()
-      .withHiveConf(hiveConf)
-      .withRetryOnException(Exception.class).build();
+      .withHiveConf(hiveConf).withFailOnException(RuntimeException.class)
+      .withRetryOnException(SemanticException.class).build();
     try {
       return retryable.executeCallable(() -> importRangerPoliciesPlain(jsonRangerExportPolicyList,
               rangerPoliciesJsonFileName,
               serviceMapJsonFileName, jsonServiceMap, finalUrl, rangerExportPolicyList, hiveConf));
+    } catch (RuntimeException e) {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:247]: you are catching a runtime exception and throwing a semantic exception. why are you doing that?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 15]
```diff
@@ -79,32 +79,29 @@
...
     if (StringUtils.isEmpty(rangerHiveServiceName)) {
       throw new SemanticException(ErrorMsg.REPL_INVALID_CONFIG_FOR_SERVICE.format("Ranger Service Name " +
         "cannot be empty", ReplUtils.REPL_RANGER_SERVICE));
     }
+    String finalUrl = getRangerExportUrl(sourceRangerEndpoint, rangerHiveServiceName, dbName);
+    LOG.debug("Url to export policies from source Ranger: {}", finalUrl);
     Retryable retryable = Retryable.builder()
       .withHiveConf(hiveConf).withFailOnException(RuntimeException.class)
-      .withRetryOnException(URISyntaxException.class).build();
+      .withRetryOnException(Exception.class).build();
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:92]: RuntimeException is a subclass of Exception. If you retry on Exception, RuntimeException will also get retried

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java [hunk 16]
```diff
@@ -79,34 +74,34 @@
   private static final String RANGER_REST_URL_EXPORTJSONFILE = "service/plugins/policies/exportJson";
   private static final String RANGER_REST_URL_IMPORTJSONFILE =
       "service/plugins/policies/importPoliciesFromFile";
+  private static final String RANGER_REST_URL_DELETEPOLICY = "service/public/v2/api/policy";
 
   public RangerExportPolicyList exportRangerPolicies(String sourceRangerEndpoint,
                                                      String dbName,
                                                      String rangerHiveServiceName,
-                                                     HiveConf hiveConf)throws SemanticException {
+                                                     HiveConf hiveConf)throws Exception {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:82]: why is this changed to exception?

Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java:82]: Now, it can throw RunTime Exception also.

**Review by aasha — APPROVED** [review]
